mqtt/mqtt.js

```diff
@@ -87,125 +57,77 @@ nats.subscribe('channel.*', function (msg) {
  */
 // AuthZ PUB
 aedes.authorizePublish = function (client, packet, callback) {
-    // Topics are in the form `mainflux/channels/<channel_id>/messages/senml-json`
-    var channel = packet.topic.split('/')[2];
-
-    /**
-     * Check if PUB is authorized
-     */
-    var options = {
-        url: config.auth_url + ':' + config.auth_port + '/channels/' + channel + '/access-grant',
-        method: 'GET',
-        headers: {
-            'Content-Type': 'application/json',
-            'Authorization': client.password
-        }
-    };
-
-    request(options, function (err, res) {
-        var error = null;
-        var msg = {};
-        if (res && (res.statusCode === 200)) {
-            console.log('Publish authorized OK');
-
-          var msg = message.RawMessage.encode({
-              /**
-               * We must publish on NATS here, because on_publish() is also called
-               * when we receive message from NATS from other adapters (in nats.subscribe()),
-               * so we must avoid re-publishing on NATS what came from other adapters
-               */
-              Publisher: client.id,
-              Channel: channel,
-              Protocol: 'mqtt',
-              ContentType: packet.topic.split('/')[4],
-              Payload: packet.payload
+    // Topics are in the form `channels/<channel_id>/messages/senml-json`
+    var channel = packet.topic.split('/')[1];
+
+    things.CanAccess({
+        token: client.password,
+        chanID: channel
+    }, function (err, res) {
+        if (!err) {
+            logger.info('authorized publish');
+            /**
+             * We must publish on NATS here, because on_publish() is also called
+             * when we receive message from NATS from other adapters (in nats.subscribe()),
+             * so we must avoid re-publishing on NATS what came from other adapters
+             */
+            var rawMsg = message.RawMessage.encode({
+                Publisher: client.id,
+                Channel: channel,
+                Protocol: 'mqtt',
+                ContentType: packet.topic.split('/')[3],
+                Payload: packet.payload
             });
 
-            console.log(msg);
-
-            console.log(util.inspect(packet, false, null))
-
             // Pub on NATS
-            nats.publish('channel.' + channel, msg);
+            nats.publish('channel.' + channel, rawMsg);
         } else {
-            console.log('Publish not authorized');
-            error = 4; // Bad username or password
+            logger.warn("unauthorized publish: %s", err.message);
+            callback(4); // Bad username or password
```

Comment:
Moving callback here (in case of error), you are not calling the callback in case there is no error.

I am not sure, but I think that it is expected to call `callback(0)` in that case - please check the doc.

Reply:
I am not calling `callback` here on purpose. Message will be sent two times over MQTT if I call `callback` here. `callback` call is not necessary in this case.

Reply:
So if I understand well - you are avoiding message to be published over MQTT by not calling this callback - you are just publishing to NATS and not to the other MQTT clients connected to MQTT broker.

Then message comes back over NATS and is published to other clients.

If this is correct - how does this not calling callback affects this NodeJS server (because obviously this is not the way that author imagined the use). @nmarcetic can there be some unexpected issues when not calling this callback (unreleased resources, memory leak, something other)?

Reply:
@nmarcetic @anovakovic01 any opinions on this?

Reply:
You are right, this callback should be called, in order to send `PUBACK` to publisher.